sympy/matrices/matrices.py

```diff
@@ -3966,6 +3966,72 @@ def vech(self, diagonal=True, check_symmetry=True):
                     count += 1
         return v
 
+    def upper(a, diag=0):
+        """
+        Examples
+        ========
+
+        >>> from sympy import Matrix, ones
+        >>> Matrix.upper(ones(4,2))
+        Matrix([
+        [1, 1],
+        [0, 1],
+        [0, 0],
+        [0, 0]])
+        >>> Matrix.upper(ones(2,4))
+        Matrix([
+        [1, 1, 1, 1],
+        [0, 1, 1, 1]])
+        >>> Matrix.upper(ones(3), diag=1)
+        Matrix([
+        [0, 1, 1],
+        [0, 0, 1],
+        [0, 0, 0]])
+        """
+        from sympy import Matrix
+
+        b = Matrix(a)
+        d = min(a.rows, a.cols)
+        for i in range(d):
+            for j in range(i + diag):
+                b[i, j] = 0
+        for i in range(d, a.rows):
+            for j in range(a.cols):
+                b[i, j] = 0
+        return b
```

Comment:
You should recast the matrix back to the type of `self` (dense vs sparse, mutable vs immutable)

Reply:
OkaI will do it.

Reply:
Rather than using a mutable matrix and making assignments, you can do this directly by passing a function to the matrix constructor. e.g. `type(self)(self.rows, self.cols, lambda i,j: self[i,j] if i < j else 0)` or something similar.